### ISSUE TITLE
Config value event interpolation

### ISSUE THREAD

**Issue by binarylogic** [vectordotdev/vector#522]
**Note: please see https://github.com/timberio/vector/issues/522#issuecomment-508219564 before starting work**

This issue outlines an interpolation syntax for select configuration values with the goal of dynamically interpolating variables from individual event data.

## Use Case

The target use case for this feature is dynamic partitioning, specifically the `aws_cloud_watch_logs` sink. This sink takes `group_name` and `stream_name` options. Both of these options should be dynamic based on data in each event, creating dynamic partitions.

For example, the default behavior of the CW agent is create a `group_name` for each log path (`/var/log/nginx.log`) and a `stream_name` representing the instance ID (`i-fds32h234hj324`). 

## Proposed Solution

In order to achieve the above we must support an event interpolation syntax for the `group_name` and `stream_name` options. For example:

```toml
[sinks.cw_log]
  type = "aws_cloudwatch_logs"
  group_name = "${event.file}"
  stream_name = "${event.ec2.instance_id}"
  # ...
```

Note, the syntax follows our env var itnerpolation syntax which I'd like to keep consistent.

## Example

Given the following event:

```json
{
  "file": "/var/log/nginx.log",
  "ec2": {
    "instance_id": "i-fds32h234hj324"
  }
}
```

When writen to the `aws_cloudwatch_logs` sink it would use:

* `group_name` = "/var/log/nginx.log"
* `stream_name` = "i-fds32h234hj324"

## Requirements

- [ ] Allow for interpolation of any event attribute
- [ ] Support a syntax for nested attributes
- [ ] Log an error and discard the event if a specified value cannot be coerced into a string (array, table, etc). We should keep this simple to start with supporting only primitive values (string, int, float, bool).
- [ ] This feature should only be enabled and supported for these 2 specific fields only. If this syntax is used in another field then it should use it literally.

## Follow Up Features

- [ ] Event `timestamp` formatting: `${event.timestamp|format "%F"}`
- [ ] Implement this in the `aws_s3`, `elasticsearch`, `http`, and more sinks. It should be easy to mark a config value as "interpolateable".

**Comment by lukesteensen** [comment]
We already have [interpolation functionality][0] for env vars in config files that would be pretty easy to reuse. Since event fields are already flattened, I think everything would just work.

[0]: https://github.com/timberio/vector/blob/8de9ad3a22f0c4789a760b4f0e57a84163edddec/src/topology/config/vars.rs

**Comment by LucioFranco** [comment]
@lukesteensen though I believe those interpolated values are precomputed at startup, whereas for event items we need to fetch them at event encoding.

**Comment by lukesteensen** [comment]
The function I linked to interpolates variables into a given string. Right now it takes a `HashMap<String, String>` as the source of the variables, but we could easily do the same thing using an event's fields as the source of the variables.

**Comment by lukesteensen** [comment]
@LucioFranco and I discussed this a bit, so I'll try to summarize here.

We broke the problem down into roughly three parts:
1. Delimiters (e.g. `${...}` vs `{{...}}` vs `#{...}` etc)
2. Internal syntax (mostly `event.key` vs `key`)
3. Extra features (e.g. default values, truncation, formatting, etc)

For delimiters, we realized a big problem with reusing `${...}` is that it clashes with the existing config file env var interpolation. Since these processes operate at totally different levels, it seems important that they're easy to distinguish both for users and our code. After eliminating `${...}`, we couldn't think of any strong arguments for others beyond simple familiarity. Given that, we decided to go with `{{...}}` which matches a huge number of common templating libraries and seemed most likely to be familiar to users.

Internal syntax seemed mostly like a stylistic preference issue. The `event.key` syntax feels natural due to its similarity to normal programming. On the other hand, that familiarity could cause confusion if it causes users to assume they're in a programming context, which they are not. The `event.` portion is also unnecessary prefix that users would have to type in every case. Since there is no other context from which to draw keys, there doesn't seem to be any reason to specify the `event` namespace. For those reasons, we went with the simple `{{ key }}` syntax.

As far as extra features, we decided to punt for the time being. We expect to be able to add any of them we need in the future, in a backwards compatible fashion. If/when we do, we liked the idea of a Jinja-like `key | filter` syntax.

An initial version is implemented for cloudwatch only as part of #519, so the next step will be to extract that to a common implementation and start using it in other contexts.
